receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go

```diff
@@ -0,0 +1,435 @@
+	if ok2 {
+		return tsm2
+	}
+	tsm2 = newTimeseriesMapPdata()
+	jm.jobsMap[sig] = tsm2
+	return tsm2
+}
+
+// MetricsAdjusterPdata takes a map from a metric instance to the initial point in the metrics instance
+// and provides AdjustMetrics, which takes a sequence of metrics and adjust their start times based on
+// the initial points.
+type MetricsAdjusterPdata struct {
+	tsm    *timeseriesMapPdata
+	logger *zap.Logger
+}
+
+// NewMetricsAdjuster is a constructor for MetricsAdjuster.
+func NewMetricsAdjusterPdata(tsm *timeseriesMapPdata, logger *zap.Logger) *MetricsAdjusterPdata {
+	return &MetricsAdjusterPdata{
+		tsm:    tsm,
+		logger: logger,
+	}
+}
+
+// AdjustMetrics takes a sequence of metrics and adjust their start times based on the initial and
+// previous points in the timeseriesMap.
+// Returns the total number of timeseries that had reset start times.
+func (ma *MetricsAdjusterPdata) AdjustMetrics(metricL *pdata.MetricSlice) int {
+	resets := 0
+	ma.tsm.Lock()
+	defer ma.tsm.Unlock()
+	for i := 0; i < metricL.Len(); i++ {
+		metric := metricL.At(i)
+		resets += ma.adjustMetric(&metric)
+	}
+	return resets
+}
+
+// Returns the number of timeseries with reset start times.
+func (ma *MetricsAdjusterPdata) adjustMetric(metric *pdata.Metric) int {
+	switch metric.DataType() {
+	case pdata.MetricDataTypeGauge:
+		// gauges don't need to be adjusted so no additional processing is necessary
+		return 0
+	default:
+		return ma.adjustMetricPoints(metric)
+	}
+}
+
+// Returns  the number of timeseries that had reset start times.
+func (ma *MetricsAdjusterPdata) adjustMetricPoints(metric *pdata.Metric) int {
+	switch dataType := metric.DataType(); dataType {
+	case pdata.MetricDataTypeGauge:
+		return ma.adjustMetricGauge(metric)
+
+	case pdata.MetricDataTypeHistogram:
+		return ma.adjustMetricHistogram(metric)
+
+	case pdata.MetricDataTypeSummary:
+		return ma.adjustMetricSummary(metric)
+
+	case pdata.MetricDataTypeSum:
+		return ma.adjustMetricSum(metric)
+
+	default:
+		// this shouldn't happen
+		ma.logger.Info("Adjust - skipping unexpected point", zap.String("type", dataType.String()))
+		return 0
+	}
+}
+
+// Returns true if 'current' was adjusted and false if 'current' is an the initial occurrence or a
+// reset of the timeseries.
+func (ma *MetricsAdjusterPdata) adjustMetricGauge(current *pdata.Metric) (resets int) {
+	currentPoints := current.Gauge().DataPoints()
+
+	for i := 0; i < currentPoints.Len(); i++ {
+		currentGauge := currentPoints.At(i)
+		tsi := ma.tsm.get(current, currentGauge.Attributes())
+		previous := tsi.previous
+		tsi.previous = current
+		if tsi.initial == nil {
+			// initial || reset timeseries.
+			tsi.initial = current
+			resets++
+			continue
+		}
+		initialPoints := tsi.initial.Gauge().DataPoints()
+		previousPoints := previous.Gauge().DataPoints()
+		if i >= initialPoints.Len() || i >= previousPoints.Len() {
+			ma.logger.Info("Adjusting Points, all lengths should be equal",
```

Comment:
Should this be debug?

Reply:
Seems reasonable.  I think this generally shouldn't happen, but isn't really actionable either and thus makes sense as a debug log.